### PR TITLE
feature: adding a log interface for error logging

### PR DESCRIPTION
**Pull request by shairozan**
* If provided (can be nil) will log any errors during usecase middleware before returning or after execution of the UseCaseFunc before returning the error. 
* Log package contains interface as well as two simple wrappers for native log and logrus

### CODE REVIEW

File: example/nodes/dog/dog.go [hunk 1]
```diff
@@ -12,12 +13,12 @@ import (
 func New(service *web.Service, logger *log.Logger) (*node.Node, error) {
 	var dogUseCase usecase.UseCase[usecase2.DogWalkRequest, *usecase2.DogWalkResponse]
 	var err error
-	if dogUseCase, err = usecase2.MakeDogWalkUseCase(logger); err != nil {
+	if dogUseCase, err = usecase2.MakeDogWalkUseCase(logger, log2.NewLogWrapper(logger)); err != nil {
```
Comment (shairozan) [inline example/nodes/dog/dog.go:16]: Most of these are for the example project.

File: log/log.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package log
+
+import (
+	"github.com/sirupsen/logrus"
+	"log"
+)
+
+type UseCaseLogger interface {
+	Log(args ...any)
```
Comment (shairozan) [inline log/log.go:9]: Figured I'd leave this interface as bland as possible. The wrappers themselves can add decoration etc

File: usecase.go [hunk 3]
```diff
@@ -82,7 +93,14 @@ func (i UseCase[I, O]) Interactor() usecase.Interactor {
 	return u
 }
 
-func New[I any, O any](input I, output O, interactor UseCaseFunc[I, O], decorationFunc func(IOInteractor *usecase.IOInteractor), m ...Middleware[I, O]) (UseCase[I, O], error) {
+func New[I any, O any](input I,
+	output O,
+	interactor UseCaseFunc[I, O],
+	decorationFunc func(IOInteractor *usecase.IOInteractor),
+	logger log.UseCaseLogger,
```
Comment (elisarver) [inline usecase.go:100]: there's a `logging.Logger` interface which you could wrap for your `UseCaseLogger`

Comment (shairozan) [reply usecase.go:100]: What package is that referring to? I can't seem to find logging.Logger or `log.Logger`?

Comment (elisarver) [reply usecase.go:100]: could have sworn there was one in there

**Review by elisarver — APPROVED** [review]
couldn't find approve in the UI, so I restarted it

**Pull request merged**